### PR TITLE
HTBHF-1479 Make sure that a payment is saved for both first payments and normal payments should any failure in the event be detected.

### PR DESCRIPTION
**Pull request by mikeSmith08**

### CODE REVIEW

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/MessageTypeProcessor.java [hunk 1]
```diff
@@ -22,4 +23,15 @@
...
     MessageType supportsMessageType();
 
+    /**
+     * Can be implemented by a MessageTypeProcessor to define custom functionality that needs to be performed
+     * when the processing of a message has failed. Default functionality as defined here is to do nothing.
+     *
+     * @param message      The message that has failed
+     * @param failureEvent The details of the failed event.
+     */
+    @SuppressWarnings("PMD.UncommentedEmptyMethodBody")
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/MessageTypeProcessor.java]: Its annoying I had to add this because its JavaDoc'd but PMD can't see it for some reason.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/MakePaymentMessageProcessor.java [hunk 2]
```diff
@@ -40,4 +41,11 @@ public MessageType supportsMessageType() {
         return MAKE_PAYMENT;
     }
 
+    @Override
+    @Transactional(Transactional.TxType.REQUIRES_NEW)
+    public void processFailedMessage(Message message, FailureEvent failureEvent) {
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/MakePaymentMessageProcessor.java]: This (and its test) are identical in both payment processors so I was wondering whether they should have some shared inheritance. Any thoughts? At least it should be in another PR if that's the case so I've left it for now.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/service/payments/PaymentService.java [hunk 3]
```diff
@@ -34,6 +40,35 @@
...
+                    .claim(paymentCycle.getClaim())
+                    .paymentAmountInPence(amountToPayInPence)
+                    .paymentCycle(paymentCycle)
+                    .paymentReference(paymentReference)
+                    .paymentStatus(PaymentStatus.FAILURE)
+                    .build();
+
+            paymentRepository.save(failedPayment);
+        } catch (Exception e) {
+            log.error("Unexpected exception caught saving a failed payment for paymentCycle: {}, cardAccountId: {}, failureEvent: {}, exception detail: {}",
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/service/payments/PaymentService.java]: I didn't see what else we could do should this attempt to save the failed payment also fail.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/MessageTypeProcessor.java (hunk 1 above)
Comment (YetAnotherMatt) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/MessageTypeProcessor.java]: We can remove the `@SuppressWarnings` and add a `// No action required` comment in the method body itself.

File: api/src/main/resources/application.yml [hunk 4]
```diff
@@ -72,7 +72,7 @@ payment-cycle:
     weeks-after-due-date: 8
   schedule:
     # run at half-past the hour from 02:30 TO 05:30
-    cron-schedule: "0 30 2-5 * * *"
+    cron-schedule: "30 0 * * * *"
```
Comment (YetAnotherMatt) [inline api/src/main/resources/application.yml]: Is this 30 seconds past each hour?

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/payments/PaymentServiceTest.java [hunk 5]
```diff
@@ -48,32 +54,48 @@
 import static uk.gov.dhsc.htbhf.logging.event.FailureEvent.FAILED_EVENT_KEY;
 import static uk.gov.dhsc.htbhf.logging.event.FailureEvent.FAILURE_DESCRIPTION_KEY;
 
-@ExtendWith(MockitoExtension.class)
+@ExtendWith(SpringExtension.class)
+@SpringBootTest
```
Comment (YetAnotherMatt) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/payments/PaymentServiceTest.java]: Remind me why this needs to be a SpringBootTest? In general we need to reduce the number of SpringBootTests, as our build times are getting dangerously long.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/MessageTypeProcessor.java (hunk 1 above)
Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/MessageTypeProcessor.java]: Removed suppress warnings.

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/payments/PaymentServiceTest.java (hunk 5 above)
Comment (dwybourn) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/payments/PaymentServiceTest.java]: So we know that we're successfully creating entities in a valid state before persisting them.

File: api/src/main/resources/application.yml (hunk 4 above)
Comment (dwybourn) [reply api/src/main/resources/application.yml]: Accidental check in, now reverted.
